include/openssl/ess.h

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef HEADER_ESS_H
+# define HEADER_ESS_H
+
+# include <openssl/opensslconf.h>
+
+# ifndef OPENSSL_NO_CMS
```

Comment:
Please remove this and let the ESS stuff be defined at all times

OR

Change it to this:

``` C
# if !defined(OPENSSL_NO_CMS) || !defined(OPENSSL_NO_TS)
```

This ensures that the ESS stuff isn't unavailable to TS when CMS is disabled.

Reply:
Changing that line will require that you do another `make update`.  You will see it have an effect on `util/libcrypto.num`

Reply:
> 
> 
> Please remove this and let the ESS stuff be defined at all times
> 
> OR
> 
> Change it to this:
> 
> ```c
> # if !defined(OPENSSL_NO_CMS) || !defined(OPENSSL_NO_TS)
> ```
> 
> This ensures that the ESS stuff isn't unavailable to TS when CMS is disabled.

AFAIK, TS can't work without CMS.
Moreover, if this constraint is not cheked/enforced in MakeFile.

Reply:
> AFAIK, TS can't work without CMS.

@levitte If this is a general dependency, shouldn’t it be handled by the configure skript instead?

Reply:
> @levitte If this is a general dependency, shouldn’t it be handled by the configure script instead?

ping @levitte

Reply:
> AFAIK, TS can't work without CMS.

Depends what you mean.  TS isn't depending on CMS per se, but its use as time stamp token in a CMS message does of course depend on CMS.  You'll find the little bit said about it in RFC 3161 [Appendix A](https://tools.ietf.org/html/rfc3161#page-20)

Also, something to be noted is that our TS code doesn't use any CMS calls at all.  Only PKCS7 calls, and some SMIME nids.

I'm unsure at this point where ESS falls in among these dependencies.  Is ESS only to be used with CMS?

Reply:
To be honest, I don't know. I was only the helping gitechnician. @opensignature and/or @FdaSilvaYY can tell you more.

Reply:
This brings us a kind of `chicken and egg` problem: how to share the part of this code between the ts and cms , knowing that you may disable any of the two ? 
New separate folder, move it to asn1 or pkcs7 folder ?

Reply:
Disclaimer: I have not read up on ESS

The ESS functions look like support functions that do not depend on CMS or TS, so they should be collected in some other folder...  I cannot really say if that should be `crypto/asn1` or `crypto/x509` (the later because there seems to be some relationship between ESS and X.509, but I've no idea how deeply that goes).

So I suggest doing exactly that, and only keeping the CMS and TS specific stuff in `crypto/cms` and `crypto/ts`